### DISCUSSION TITLE
ivadomed and PSIR (phase-sensitive inversion recovery) images

### DISCUSSION THREAD

**Discussion by valosekj** [ivadomed/ivadomed#1165] [Q&A]
Dear ivadomed team,

I would like to train segmentation models on PSIR (phase-sensitive inversion recovery) images. Since it looks that [PSIR is not supported by BIDS](https://neurostars.org/t/bids-naming-for-the-psir-sequence/16592), I tried to use `IRT1` as a suffix instead of `PSIR`. However, when I ran `ivadomed -c config.json` with:

```
...
        "contrast_params": {
            "training_validation": ["IRT1"],
            "testing": ["IRT1"],
            "balance": {}
        },
...
```

I got the following error:

```
...
2022-06-06 10:21:43.424 | WARNING  | ivadomed.loader.bids_dataframe:create_bids_dataframe:139 - No subject files were found in 'bids_seg_sc_psir' dataset. Skipping dataset.
...
RuntimeError: No subject files found. Check selection of parameters in config.json and datasets compliance with BIDS specification.  
```

If I replaced `IRT1` with, for example, `T2w` in filenames and config.json file, model training started successfully. 
Is this the proper way to deal with PSIR images, or is there any other recommendation how to rename them to be compatible with both BIDS and ivadomed?

Thanks,
Jan

**Comment by jcohenadad** [comment]
@valosekj thank you for reaching out. I think the suggestion in the answer to my original post on neurostars had some issues, so I've updated it: [1m](https://neurostars.org/t/bids-naming-for-the-psir-sequence/16592/3?u=jcohenadad)

Regardless of this minor semantic issue, I'm curious as to why IRT1 fails in your case. Given that BEP01 was merged 'relatively' recently, I am wondering if your version of ivadomed includes a recent-enough pybids version that recognizes IRT1. Is there a way to 'easily' test this hypothesis @ivadomed-team?

**Comment by dyt811** [comment]
Hi, @valosekj, thank you for reporting this issue to us. Do you happen to have the folder structure (or tree structure) of the data files organizations (including /derivatves), showing how they are arranged? The data may gets excluded/not indexed if no proper derivatives are provided. Could you provide the full JSON along with the data files organization so we can try to replicate and investigate this deeper? Thanks.

**Comment by mariehbourget** [comment]
> Regardless of this minor semantic issue, I'm curious as to why IRT1 fails in your case. Given that BEP01 was merged 'relatively' recently, I am wondering if your version of ivadomed includes a recent-enough pybids version that recognizes IRT1. Is there a way to 'easily' test this hypothesis @ivadomed-team?

We use the latest version of pybids, which in turn uses the latest version of the validator.
It looks like a validator issue.

**EDIT: The validator passes with the `IRT1` suffix as long as the `inv-` entity is also in the filename** (required as per the template here: https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#anatomy-imaging-data) 

~Quick test with the [validator](https://bids-standard.github.io/bids-validator/) gives me this error when using `IRT1` suffix:~
```
Error 1: [Code 1] NOT_INCLUDED
Files with such naming scheme are not part of BIDS specification. This error is most commonly caused by typos in file names
 that make them not BIDS compatible. Please consult the specification and make sure your files are named correctly.
If this is not a file naming issue (for example when including files not yet covered by the BIDS specification) you should include
a ".bidsignore" file in your dataset (see https://github.com/bids-standard/bids-validator#bidsignore for details).
Please note that derived (processed) data should be placed in /derivatives folder and source data
(such as DICOMS or behavioural logs in proprietary formats) should be placed in the /sourcedata folder.
```

~So it looks like, even though BEP001 was merged, the validator is not up-to-date.
The error message suggests to use a `.bidsignore` file in the dataset but I tested it and although it passes the validator, pybids/ivadomed seem to ignore the files as well.~

**Comment by mariehbourget** [comment]
A possible workaround that is working on my side is to add the argument `validate=False` to the indexer here:
https://github.com/ivadomed/ivadomed/blob/d8e4603910e6aed2e4c02330c4e739dbf9f8031b/ivadomed/loader/bids_dataframe.py#L106

So replace the line with:
`indexer = pybids.BIDSLayoutIndexer(force_index=force_index, validate=False)`

**Comment by jcohenadad** [comment]
> although it passes the validator, pybids/ivadomed seem to ignore the files as well

hum... this is a problem because this issue will likely happen a lot in the future, and I would like to discourage our users from manually editing ivadomed's file as suggested in https://github.com/ivadomed/ivadomed/discussions/1165#discussioncomment-2947888 because it will induce all sorts of problems (versioning issues, etc.). Should we open an issue about it?

**Reply by mariehbourget** [reply]
I think so, I remember we talked about the possibility to add an "off" switch to the validator in the config file in other issues.

**Reply by dyt811** [reply]
Issue created at https://github.com/ivadomed/ivadomed/issues/1166.

**Comment by mariehbourget** [comment]
I amended my previous [comment](https://github.com/ivadomed/ivadomed/discussions/1165#discussioncomment-2947850).
There is no error in the validator when the filename also includes the `inv-` entity (which is required for `IRT1` suffix).

However, the issue with ivadomed persists and will require more investigation.
The [workaround](https://github.com/ivadomed/ivadomed/discussions/1165#discussioncomment-2947888) remains the same.

**Comment by valosekj** [comment]
Thank you all for your feedback!

My original dataset ([for which I got the error](https://github.com/ivadomed/ivadomed/discussions/1165#discussion-4142705)) was structured:

```
├── dataset_description.json
├── participants.tsv
├── sub-5142B
│   └── anat
│       ├── sub-5142B_acq-02_part-mag_IRT1.json
│       ├── sub-5142B_acq-02_part-mag_IRT1.nii.gz
│       ├── sub-5142B_acq-03_part-mag_IRT1.json
│       ├── sub-5142B_acq-03_part-mag_IRT1.nii.gz
│       ├── sub-5142B_acq-04_part-mag_IRT1.json
│       └── sub-5142B_acq-04_part-mag_IRT1.nii.gz
...
├── derivatives
│   ├── dataset_description.json
│   └── labels
│       ├── sub-5142B
│       │   └── anat
│       │       ├── sub-5142B_acq-02_part-mag_IRT1_seg.nii.gz
│       │       ├── sub-5142B_acq-03_part-mag_IRT1_seg.nii.gz
│       │       └── sub-5142B_acq-04_part-mag_IRT1_seg.nii.gz
...

```

A few notes:
- I used only magnitude reconstruction (`part-mag`) for ivadomed training (therefore, I got rid of `part-phase` (or [`part-real`](https://neurostars.org/t/bids-naming-for-the-psir-sequence/16592/3)) reconstruction)
- I omitted the `inv` label since I had only one inversion time
- I acquired PSIR image several times for each subject (with different cardiac trigger settings); therefore, I used the `acq` label (I used `acq` over `run` since I assume that `run` should be used only when sequence settings are completely same?).

Based on @mariehbourget [suggestion](https://github.com/ivadomed/ivadomed/discussions/1165#discussioncomment-2947850), I included also `inv` label:

```
├── dataset_description.json
├── participants.tsv
├── sub-5142B
│   └── anat
│       ├── sub-5142B_acq-02_inv-01_part-mag_IRT1.json
│       ├── sub-5142B_acq-02_inv-01_part-mag_IRT1.nii.gz
│       ├── sub-5142B_acq-03_inv-01_part-mag_IRT1.json
│       ├── sub-5142B_acq-03_inv-01_part-mag_IRT1.nii.gz
│       ├── sub-5142B_acq-04_inv-01_part-mag_IRT1.json
│       └── sub-5142B_acq-04_inv-01_part-mag_IRT1.nii.gz
...
├── derivatives
│   ├── dataset_description.json
│   └── labels
│       ├── sub-5142B
│       │   └── anat
│       │       ├── sub-5142B_acq-02_inv-01_part-mag_IRT1_seg.nii.gz
│       │       ├── sub-5142B_acq-03_inv-01_part-mag_IRT1_seg.nii.gz
│       │       └── sub-5142B_acq-04_inv-01_part-mag_IRT1_seg.nii.gz
...

```

Now ivadomed accepts the dataset. So, it looks that the problem was caused by the missing `inv` label, which I intentionally omitted 😅

**Reply by mariehbourget** [reply]
Great news, thanks for the feedback @valosekj!

> * I acquired PSIR image several times for each subject (with different cardiac trigger settings); therefore, I used the `acq` label (I used `acq` over `run` since I assume that `run` should be used only when sequence settings are completely same?).

You are correct, `run-` is usually used for several acquisitions with the same settings, and `acq-` for several acquisitions with different settings.
